crates/vm/src/system/native_adapter/mod.rs

```diff
@@ -214,7 +180,34 @@ where
         instruction: &Instruction<F>,
         adapter_row: &mut [F],
     ) -> Self::ReadData {
-        todo!("Implement read operation");
+        assert!(R <= 2);
+
+        let &Instruction { b, e, .. } = instruction;
+
+        let cols: &mut NativeAdapterCols<_, R, W> = adapter_row.borrow_mut();
+
+        let mut reads = [[F::ZERO; 1]; R];
+        if R >= 1 {
+            cols.reads_aux[0].address.pointer = b;
+            reads[0][0] = tracing_read_or_imm_native(
+                memory,
+                e.as_canonical_u32(),
+                b,
+                &mut cols.reads_aux[0].address.address_space,
+                &mut cols.reads_aux[0].read_aux,
+            );
+        }
+        if R >= 2 {
+            cols.reads_aux[0].address.pointer = b;
```

Comment:
```suggestion
            cols.reads_aux[1].address.pointer = c;
```

was it supposed to be like this?